src/mpid/ch4/shm/src/shm_p2p.h

```diff
@@ -14,9 +14,84 @@
 #ifndef SHM_P2P_H_INCLUDED
 #define SHM_P2P_H_INCLUDED
 
+/*
+=== BEGIN_MPI_T_CVAR_INFO_BLOCK ===
+
+cvars:
+    - name        : MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE
+      category    : CH4
+      type        : int
+      default     : 4096
+      class       : device
+      verbosity   : MPI_T_VERBOSITY_USER_BASIC
+      scope       : MPI_T_SCOPE_ALL_EQ
+      description : >-
+        If a send message size is larger than MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE (in bytes),
+        then enable XPMEM-based single copy protocol for intranode communication. The
+        environment variable is valid only when then XPMEM shmmod is enabled.
+
+=== END_MPI_T_CVAR_INFO_BLOCK ===
+*/
+
 #include <shm.h>
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+#include "../xpmem/shm_inline.h"
+#endif
 #include "../posix/shm_inline.h"
 
+#define MPIDI_SHM_PT2PT_DEFAULT 1
+#define MPIDI_SHM_PT2PT_MULTIMODS 2
+
+/* Enable multi-shmmods protocol when more than one shmmod is enabled. */
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+#define MPIDI_SHM_PT2PT_PROT MPIDI_SHM_PT2PT_MULTIMODS
+#else
+#define MPIDI_SHM_PT2PT_PROT MPIDI_SHM_PT2PT_DEFAULT
+#endif
+
+#if (MPIDI_SHM_PT2PT_PROT == MPIDI_SHM_PT2PT_MULTIMODS)
+#undef FUNCNAME
+#define FUNCNAME MPIDI_SHM_mmods_try_matched_recv
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
+MPL_STATIC_INLINE_PREFIX int MPIDI_SHM_mmods_try_matched_recv(void *buf,
+                                                              MPI_Aint count,
+                                                              MPI_Datatype datatype,
+                                                              MPIR_Request * message,
+                                                              bool * recvd_flag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_SHM_MMODS_TRY_MATCHED_RECV);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_SHM_MMODS_TRY_MATCHED_RECV);
+
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+    /* XPMEM special receive */
+    if (MPIDI_SHM_REQUEST(message, status) & MPIDI_SHM_REQ_XPMEM_SEND_LMT) {
+        /* Matching XPMEM LMT receive is now posted */
+        MPIR_Datatype_add_ref_if_not_builtin(datatype); /* will -1 once completed in handle_lmt_recv */
+        MPIDIG_REQUEST(message, datatype) = datatype;
+        MPIDIG_REQUEST(message, buffer) = (char *) buf;
+        MPIDIG_REQUEST(message, count) = count;
+
+        MPIDI_XPMEM_am_unexp_rreq_t *unexp_rreq = &MPIDI_XPMEM_REQUEST(message, unexp_rreq);
+        mpi_errno = MPIDI_XPMEM_handle_lmt_recv(unexp_rreq->src_offset,
+                                                unexp_rreq->data_sz, unexp_rreq->sreq_ptr,
+                                                unexp_rreq->src_lrank, message);
+        if (mpi_errno)
+            MPIR_ERR_POP(mpi_errno);
+
+        *recvd_flag = true;
+    }
+#endif /* end of MPIDI_CH4_SHM_ENABLE_XPMEM */
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_SHM_MMODS_TRY_MATCHED_RECV);
+    return mpi_errno;
+  fn_fail:
```

Comment:
There's a bunch of unused label warnings if XPMEM is turned off for all of these functions. Can you add this before each of the labels?

    #if (MPIDI_SHM_PT2PT_PROT == MPIDI_SHM_PT2PT_MULTIMODS) && defined(MPIDI_CH4_SHM_ENABLE_XPMEM)

Reply:
Added